extension/entitystore/extension.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+	"github.com/jellydator/ttlcache/v3"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/extension"
+	"go.uber.org/zap"
+
+	configaws "github.com/aws/amazon-cloudwatch-agent/cfg/aws"
+	"github.com/aws/amazon-cloudwatch-agent/internal/ec2metadataprovider"
+	"github.com/aws/amazon-cloudwatch-agent/internal/retryer"
+	"github.com/aws/amazon-cloudwatch-agent/sdk/service/cloudwatchlogs"
+	"github.com/aws/amazon-cloudwatch-agent/translator/config"
+)
+
+const (
+	Service                     = "Service"
+	InstanceIDKey               = "EC2.InstanceId"
+	ASGKey                      = "EC2.AutoScalingGroup"
+	ServiceNameSourceKey        = "AWS.ServiceNameSource"
+	PlatformType                = "PlatformType"
+	EC2PlatForm                 = "AWS::EC2"
+	Type                        = "Type"
+	Name                        = "Name"
+	Environment                 = "Environment"
+	podTerminationCheckInterval = 5 * time.Minute
+)
+
+type ec2ProviderType func(string, *configaws.CredentialConfig) ec2iface.EC2API
+
+type serviceProviderInterface interface {
+	startServiceProvider()
+	addEntryForLogFile(LogFileGlob, ServiceAttribute)
+	addEntryForLogGroup(LogGroupName, ServiceAttribute)
+	logFileServiceAttribute(LogFileGlob, LogGroupName) ServiceAttribute
+	getServiceNameAndSource() (string, string)
+}
+
+type EntityStore struct {
+	logger *zap.Logger
+	config *Config
+	done   chan struct{}
+
+	// mode should be EC2, ECS, EKS, and K8S
+	mode string
+
+	kubernetesMode string
+
+	// ec2Info stores information about EC2 instances such as instance ID and
+	// auto scaling groups
+	ec2Info EC2Info
+
+	// eksInfo stores information about EKS such as pod to service Env map
+	eksInfo *eksInfo
+
+	// serviceprovider stores information about possible service names
+	// that we can attach to the entity
+	serviceprovider serviceProviderInterface
+
+	// nativeCredential stores the credential config for agent's native
+	// component such as LogAgent
+	nativeCredential client.ConfigProvider
+
+	metadataprovider ec2metadataprovider.MetadataProvider
+
+	stsClient stsiface.STSAPI
+
+	podTerminationCheckInterval time.Duration
+}
+
+var _ extension.Extension = (*EntityStore)(nil)
+
+func (e *EntityStore) Start(ctx context.Context, host component.Host) error {
+	// Get IMDS client and EC2 API client which requires region for authentication
+	// These will be passed down to any object that requires access to IMDS or EC2
+	// API client so we have single source of truth for credential
+	e.done = make(chan struct{})
+	e.metadataprovider = getMetaDataProvider()
+	e.mode = e.config.Mode
+	e.kubernetesMode = e.config.KubernetesMode
+	e.podTerminationCheckInterval = podTerminationCheckInterval
+	ec2CredentialConfig := &configaws.CredentialConfig{
+		Profile:  e.config.Profile,
+		Filename: e.config.Filename,
+	}
+	switch e.mode {
```

Comment:
If `onPrem`, only `serviceprovider` component will run. I think we still need to enable the extension for `onPrem` since there is support to associate AppSignal with LogAgent log groups which requires this extension running

Reply:
Blocking comment.

Adding a flag to ensure only IMDS calling functions will run.